keras/preprocessing/image.py

```diff
@@ -991,22 +1036,36 @@ def _list_valid_filenames_in_directory(directory, white_list_formats,
     def _recursive_list(subpath):
         return sorted(os.walk(subpath, followlinks=follow_links), key=lambda tpl: tpl[0])
 
+    def _calculate_hash_percent(filename):
+        hash_name = hashlib.sha1(as_bytes(filename)).hexdigest()
+        return ((int(hash_name, 16) % (MAX_NUM_IMAGES_PER_CLASS + 1)) *
+                (100.0 / MAX_NUM_IMAGES_PER_CLASS))
+
+    is_training = subset is not None and subset == 'training'
+    dirname = os.path.basename(directory)
+
     classes = []
     filenames = []
-    subdir = os.path.basename(directory)
-    basedir = os.path.dirname(directory)
     for root, _, files in _recursive_list(directory):
-        for fname in sorted(files):
+        for fname in files:
```

Comment:
You can make the whole PR a lot simpler by using: `for fname in sorted(files)[start : stop]`

Remember: this feature should only total a few lines of change. The only "logic" required by this PR is the computation of `start` and `stop` above given the `validation_split` passed by the user (if any).

Reply:
@fchollet just for clarification, I guess you're saying I should do away with the filename hashing for the split and just use `split=(start, stop)` ? My current implementation of filename hashing already makes the validation split consistent without having to specify the segment.

Also would `stop` be necessary or can we just treat the `validation_split` as the length? E.g. `validation_split=0.3` and `split_start=0.7` would mean treat last 30% of each directory as validation.

Reply:
`start`/`stop` is flexible since it means `_count/list_valid_files_in_directory` will be agnostic to the validation/training story, and only needs the `split` argument. But it doesn't make a big difference.

Reply:
Alright, I'll get rid of the filename hashing in favor of `(start, stop)`. So would the API would look like:
```
generator = image.ImageDataGenerator()
train_iterator = generator.flow_from_directory(image_dir, subset='training', split=(0.0, 0.4))
val_iterator = generator.flow_from_directory(image_dir, subset='validation', split=(0.4, 1.0))

train_iterator = generator.flow(x, y, subset='training', split=(0.0, 0.4))
val_iterator = generator.flow(x, y, subset='validation', split=(0.4, 1.0))
```

Reply:
`split` can be reserved for internal methods, as it isn't very user-friendly. For user-facing APIs, we can go with your original proposal:

```python
generator = image.ImageDataGenerator(validation_split=0.5)
train_iterator = generator.flow_from_directory(image_dir, subset='training')
val_iterator = generator.flow_from_directory(image_dir, subset='validation')

# This should also be possible
train_iterator = generator.flow(x, y, subset='training')
val_iterator = generator.flow(x, y, subset='validation')
```

Reply:
Thanks. Hope to fix by end of this week.

Reply:
The start/stop usage is quite useful in that you can do cross fold validation very easily with it, but I agree that it isn't very user-friendly. If it was still on the user api side it would be nice though.
Also thanks @ksindi for this update, I have also used it and it makes a lot of things easier.